File: java21/src/main/java/io/github/sps23/typeclasses/ShowExamples.java
package io.github.sps23.typeclasses;

import java.util.List;
import java.util.Map;

/**
 * Examples demonstrating different versions of the typeclass pattern in Java.
 *
 * This corresponds to the blog post "Typeclasses in Java: Because Why Should
 * Scala Have All the Fun?"
 */
public class ShowExamples {

    public static void main(String[] args) {
        version1BasicImplementation();
        System.out.println();

        version2RegistryBased();
        System.out.println();

        version4ExtensionMethods();
        System.out.println();

        version5AutoDerivation();
    }

    /**
     * Version 1: Basic typeclass implementation with explicit instance passing.
     */
    private static void version1BasicImplementation() {
        System.out.println("=== Version 1: Basic Implementation ===");

        // Create Show instances
        Show<Integer> intShow = Show.forInteger();
        Show<String> stringShow = Show.forString();

        // Use them explicitly
        Show.print(42, intShow); // Int(42)
        Show.print("hello", stringShow); // "hello"

        // Inline lambda
        Show.print(3.14, d -> "Double(" + d + ")"); // Double(3.14)

        // Show for collections
        Show<List<Integer>> listShow = Show.forList(intShow);
        Show.print(List.of(1, 2, 3), listShow); // List(Int(1), Int(2), Int(3))

        // Show for maps
        Show<Map<String, Integer>> mapShow = Show.forMap(stringShow, intShow);
        Show.print(Map.of("a", 1, "b", 2), mapShow); // Map("a" -> Int(1), "b" -> Int(2))
    }

    /**
     * Version 2: Registry-based lookup (simulates implicit resolution).
     */
    private static void version2RegistryBased() {
        System.out.println("=== Version 2: Registry-Based Lookup ===");

        // Use registered instances
        Show.print(42, Show.Registry.summon(Integer.class)); // Int(42)
        Show.print("hello", Show.Registry.summon(String.class)); // "hello"
        Show.print(true, Show.Registry.summon(Boolean.class)); // True

        // Register custom instance
        record Person(String name, int age) {
        }
        Show.Registry.register(Person.class, p -> "Person(" + p.name() + ", " + p.age() + ")");

        Person alice = new Person("Alice", 30);
        Show.print(alice, Show.Registry.summon(Person.class)); // Person(Alice, 30)
    }

    /**
     * Version 4: Extension-method-like syntax with Showable wrapper.
     */
    private static void version4ExtensionMethods() {
        System.out.println("=== Version 4: Extension Methods ===");

        // Fluent API
        Showable.of(42, Show.forInteger()).inspect() // Inspecting: Int(42)
                .map(i -> i * 2, Show.forInteger()).print(); // Int(84)

        // With registry lookup
        Showable.of("hello", String.class).inspect() // Inspecting: "hello"
                .map(String::toUpperCase, Show.forString()).print(); // "HELLO"
    }

    /**
     * Version 5: Auto-derivation for Record types.
     */
    private static void version5AutoDerivation() {
        System.out.println("=== Version 5: Auto-Derivation for Records ===");

        // Define record types
        record Person(String name, int age) {
        }
        record Address(String street, String city) {
        }
        record Employee(Person person, Address address, double salary) {
        }

        // Derive Show instances automatically
        Show<Person> personShow = ShowDerivation.deriveRecord();
        Show<Address> addressShow = ShowDerivation.deriveRecord();
        Show<Employee> employeeShow = ShowDerivation.deriveRecord();

        // Use them
        Person bob = new Person("Bob", 35);
        Address addr = new Address("123 Main St", "Springfield");
        Employee emp = new Employee(bob, addr, 75000.0);

        personShow.show(bob); // Person(name = "Bob", age = 35)
        Show.print(bob, personShow);

        addressShow.show(addr); // Address(street = "123 Main St", city = "Springfield")
        Show.print(addr, addressShow);

        employeeShow.show(emp); // Nested records work too!
        Show.print(emp, employeeShow);
    }
}

File: scala2/src/main/scala/io/github/sps23/typeclasses/ShowExamples.scala
package io.github.sps23.typeclasses

/** Examples demonstrating the typeclass pattern in Scala 2.
  *
  * This corresponds to the blog post "Typeclasses in Java: Because Why Should Scala Have All the
  * Fun?"
  */
object ShowExamples extends App {

  import Show._

  println("=== Scala 2 Typeclass Examples ===\n")

  // ============================================================================
  // Basic usage with implicit resolution
  // ============================================================================

  println("--- Basic Usage ---")

  // Compiler finds the implicit Show[Int] instance
  Show.print(42) // Int(42)

  // Compiler finds the implicit Show[String] instance
  Show.print("hello") // "hello"

  // Compiler finds the implicit Show[Boolean] instance
  Show.print(true) // True

  // Using extension method syntax
  42.print      // Int(42)
  "world".print // "world"

  println()

  // ============================================================================
  // Collections with automatic derivation
  // ============================================================================

  println("--- Collections ---")

  // Compiler automatically derives Show[List[Int]]
  val numbers = List(1, 2, 3)
  Show.print(numbers) // List(Int(1), Int(2), Int(3))
  numbers.print       // List(Int(1), Int(2), Int(3))

  // Nested collections work too!
  val nested = List(List(1, 2), List(3, 4))
  Show.print(nested) // List(List(Int(1), Int(2)), List(Int(3), Int(4)))

  // Maps
  val scores = Map("Alice" -> 95, "Bob" -> 87)
  Show.print(scores) // Map("Alice" -> Int(95), "Bob" -> Int(87))

  // Options
  val maybeValue: Option[Int] = Some(42)
  Show.print(maybeValue)        // Some(Int(42))
  Show.print(None: Option[Int]) // None

  println()

  // ============================================================================
  // Custom types with implicit instances
  // ============================================================================

  println("--- Custom Types ---")

  case class Person(name: String, age: Int)

  // Define Show instance for Person
  implicit val personShow: Show[Person] = new Show[Person] {
    def show(p: Person): String = s"Person(${p.name}, ${p.age})"
  }

  val alice = Person("Alice", 30)
  Show.print(alice) // Person(Alice, 30)
  alice.print       // Person(Alice, 30)

  // Show for List[Person] is automatically derived!
  val people = List(Person("Alice", 30), Person("Bob", 35))
  Show.print(people) // List(Person(Alice, 30), Person(Bob, 35))

  println()

  // ============================================================================
  // Context bounds syntax
  // ============================================================================

  println("--- Context Bounds ---")

  // Function using context bound (syntactic sugar)
  def printAll[A: Show](items: List[A]): Unit =
    items.foreach(item => Show.print(item))

  // Same as above but more explicit
  def printAllExplicit[A](items: List[A])(implicit show: Show[A]): Unit =
    items.foreach(item => println(show.show(item)))

  printAll(List(1, 2, 3))
  printAll(List("a", "b", "c"))
  printAll(List(Person("Charlie", 40), Person("Diana", 28)))

  println()

  // ============================================================================
  // Summoning instances
  // ============================================================================

  println("--- Summoning Instances ---")

  // Get the implicit Show instance
  val intShowInstance = implicitly[Show[Int]]
  println(intShowInstance.show(99)) // Int(99)

  // Or using Show.apply
  val stringShowInstance = Show[String]
  println(stringShowInstance.show("scala")) // "scala"
}
